### DISCUSSION TITLE
What is the best way to detect whether the http server/CDN serving the PMtiles file is down via an error/event through the PMtiles library?

### DISCUSSION THREAD

**Discussion by CATALYST1109** [protomaps/PMTiles#196] [Q&A] [answered]
Firstly, thanks for creating this. The whole idea of being able to serve a map without a service is great, and saves a lot of headache !
Now on to the issue -- I was able to implement a MaplibreGl Js map with a pmtiles files being served by http-server as in the docs (from my local machine). My question is whether there exists a way to catch the uncaught error(see below) that occurs when the file is not being served (the server is down)? 
I tried using  maplibre map events like 'error','sourcedataabort' and 'dataabort', but none of them fired upon this error, hence why I'm posting here and not on  Maplibre. Maybe someway PMtiles could indicate to Maplibre that source is inaccessible so that it in turn  aborts source loading?
I can run a get request to the pmtiles server upon loading the map component and based on the response,  display an error, but I wanted to check if it could be done using existing library code, which seems cleaner.
I'm fairly new to JS, so apologies if it sounds noobish. I'm using all this in an Angular (11.x) app btw.

Error stacktrace from console :-
core.js:6210 ERROR Error: Uncaught (in promise): TypeError: Failed to fetch
TypeError: Failed to fetch
    at zone-evergreen.js:1505:1
    at proto.<computed> (zone-evergreen.js:973:1)
    at FetchSource.<anonymous> (index.mjs:1010:1)
    at Generator.next (<anonymous>)
    at index.mjs:18:1
    at new ZoneAwarePromise (zone-evergreen.js:1427:1)
    at __async (index.mjs:2:1)
    at FetchSource.getBytes (index.mjs:1004:1)
    at index.mjs:1133:1
    at Generator.next (<anonymous>)
    at resolvePromise (zone-evergreen.js:1211:1)
    at resolvePromise (zone-evergreen.js:1165:1)
    at zone-evergreen.js:1278:1
    at _ZoneDelegate.invokeTask (zone-evergreen.js:406:1)
    at Object.onInvokeTask (core.js:28578:1)
    at _ZoneDelegate.invokeTask (zone-evergreen.js:405:1)
    at Zone.runTask (zone-evergreen.js:178:1)
    at drainMicroTaskQueue (zone-evergreen.js:585:1)
defaultErrorLogger @ core.js:6210
handleError @ core.js:6258
next @ core.js:29198
__tryOrUnsub @ Subscriber.js:183
next @ Subscriber.js:122
_next @ Subscriber.js:72
next @ Subscriber.js:49
next @ Subject.js:39
emit @ core.js:25940
(anonymous) @ core.js:28617
invoke @ zone-evergreen.js:372
run @ zone-evergreen.js:134
runOutsideAngular @ core.js:28520
onHandleError @ core.js:28617
handleError @ zone-evergreen.js:376
runGuarded @ zone-evergreen.js:147
api.microtaskDrainDone @ zone-evergreen.js:1072
drainMicroTaskQueue @ zone-evergreen.js:592
Promise.then (async)
nativeScheduleMicroTask @ zone-evergreen.js:561
scheduleMicroTask @ zone-evergreen.js:572
scheduleTask @ zone-evergreen.js:396
onScheduleTask @ zone-evergreen.js:283
scheduleTask @ zone-evergreen.js:386
scheduleTask @ zone-evergreen.js:221
scheduleMicroTask @ zone-evergreen.js:241
scheduleResolveOrReject @ zone-evergreen.js:1265
resolvePromise @ zone-evergreen.js:1202
(anonymous) @ zone-evergreen.js:1118
Promise.then (async)
(anonymous) @ zone-evergreen.js:1496
ZoneAwarePromise @ zone-evergreen.js:1427
Ctor.then @ zone-evergreen.js:1495
resolvePromise @ zone-evergreen.js:1169
resolve @ zone-evergreen.js:1294
step @ index.mjs:17
(anonymous) @ index.mjs:18
ZoneAwarePromise @ zone-evergreen.js:1427
__async @ index.mjs:2
getBytes @ index.mjs:1004
(anonymous) @ index.mjs:1133
(anonymous) @ index.mjs:18
ZoneAwarePromise @ zone-evergreen.js:1427
__async @ index.mjs:2
getHeaderAndRoot @ index.mjs:1132
(anonymous) @ index.mjs:1296
ZoneAwarePromise @ zone-evergreen.js:1427
(anonymous) @ index.mjs:1295
(anonymous) @ index.mjs:18
ZoneAwarePromise @ zone-evergreen.js:1427
__async @ index.mjs:2
getHeader @ index.mjs:1288
(anonymous) @ index.mjs:1402
(anonymous) @ index.mjs:18
ZoneAwarePromise @ zone-evergreen.js:1427
__async @ index.mjs:2
getHeader @ index.mjs:1401
Protocol.tile @ index.mjs:707
T @ maplibre-gl.js:31
push.t9QN.t.getJSON @ maplibre-gl.js:31
L @ maplibre-gl.js:35
load @ maplibre-gl.js:35
onAdd @ maplibre-gl.js:35
onAdd @ maplibre-gl.js:35
addSource @ maplibre-gl.js:35
_load @ maplibre-gl.js:35
(anonymous) @ maplibre-gl.js:35
timer @ zone-evergreen.js:2378
invokeTask @ zone-evergreen.js:406
onInvokeTask @ core.js:28578
invokeTask @ zone-evergreen.js:405
runTask @ zone-evergreen.js:178
invokeTask @ zone-evergreen.js:487
ZoneTask.invoke @ zone-evergreen.js:476
data.args.<computed> @ zone-evergreen.js:2358
requestAnimationFrame (async)
scheduleTask @ zone-evergreen.js:2360
scheduleTask @ zone-evergreen.js:393
onScheduleTask @ zone-evergreen.js:283
scheduleTask @ zone-evergreen.js:386
scheduleTask @ zone-evergreen.js:221
scheduleMacroTask @ zone-evergreen.js:244
scheduleMacroTaskWithCurrentZone @ zone-evergreen.js:683
(anonymous) @ zone-evergreen.js:2402
proto.<computed> @ zone-evergreen.js:973
frame @ maplibre-gl.js:31
loadJSON @ maplibre-gl.js:35
_updateStyle @ maplibre-gl.js:35
setStyle @ maplibre-gl.js:35
Map @ maplibre-gl.js:35
ngAfterViewInit @ libre-map.component.ts:68
callHook @ core.js:2573
callHooks @ core.js:2542
executeInitAndCheckHooks @ core.js:2493
refreshView @ core.js:9551
refreshEmbeddedViews @ core.js:10605
refreshView @ core.js:9504
refreshComponent @ core.js:10651
refreshChildComponents @ core.js:9277
refreshView @ core.js:9530
renderComponentOrTemplate @ core.js:9594
tickRootContext @ core.js:10825
detectChangesInRootView @ core.js:10850
detectChanges @ core.js:22853
tick @ core.js:29572
(anonymous) @ core.js:29453
invoke @ zone-evergreen.js:372
onInvoke @ core.js:28591
invoke @ zone-evergreen.js:371
run @ zone-evergreen.js:134
run @ core.js:28475
next @ core.js:29452
__tryOrUnsub @ Subscriber.js:183
next @ Subscriber.js:122
_next @ Subscriber.js:72
next @ Subscriber.js:49
next @ Subject.js:39
emit @ core.js:25940
checkStable @ core.js:28528
onHasTask @ core.js:28608
hasTask @ zone-evergreen.js:426
_updateTaskCount @ zone-evergreen.js:447
_updateTaskCount @ zone-evergreen.js:274
runTask @ zone-evergreen.js:195
drainMicroTaskQueue @ zone-evergreen.js:585
Promise.then (async)
nativeScheduleMicroTask @ zone-evergreen.js:561
scheduleMicroTask @ zone-evergreen.js:572
scheduleTask @ zone-evergreen.js:396
scheduleTask @ zone-evergreen.js:221
scheduleMicroTask @ zone-evergreen.js:241
scheduleResolveOrReject @ zone-evergreen.js:1265
then @ zone-evergreen.js:1450
bootstrapModule @ core.js:29240
zUnb @ main.ts:11
__webpack_require__ @ bootstrap:84
0 @ main.js:11
__webpack_require__ @ bootstrap:84
checkDeferredModules @ bootstrap:45
webpackJsonpCallback @ bootstrap:32
(anonymous) @ main.js:1

**Comment by bdon** [comment, marked as answer]
Hi, you are right - any errors aren't surfaced through MapLibre correctly right now.

I opened [a PR](https://github.com/protomaps/PMTiles/pull/199) which should let you catch error events in MapLibre when any part of pmtiles loading fails, example:

```js
map.on('error', function() {
console.log('A error event occurred.');
});
```

**Reply by bdon** [reply]
@CATALYST1109 can you try pmtiles v2.8.0 with maplibre error events?

**Reply by CATALYST1109** [reply]
@bdon Thanks for jumping on this so quickly! I tried it out with 2.8.0 and can confirm that it works
